src/MongoDB.Driver.GridFS/GridFSForwardOnlyDownloadStream.cs

```diff
@@ -158,8 +145,24 @@ public override long Seek(long offset, SeekOrigin origin)
         }
 
         // protected methods
+        protected override void CloseImplementation(CancellationToken cancellationToken)
+        {
+            if (_checkMD5 && _position == FileInfo.Length)
+            {
+                var md5 = BsonUtils.ToHexString(_md5.GetHashAndReset());
+                if (!md5.Equals(FileInfo.MD5, StringComparison.OrdinalIgnoreCase))
+                {
+#pragma warning disable 618
+                    throw new GridFSMD5Exception(_idAsBsonValue);
+#pragma warning restore
+                }
+            }
+        }
+
         protected override void Dispose(bool disposing)
         {
+            CloseIfNotAlreadyClosedFromDispose(disposing);
+
```

Comment:
We have to duplicate this call in the subclass also because the Stream must be closed before any of its resources are Disposed below.